### PR TITLE
expression: implement vectorized evaluation for `builtinTidbParseTsoSig`

### PR DESCRIPTION
**Pull request by AerysNan**
Signed-off-by: AerysNan <aerysnan@gmail.com>

PCP #12101

### What problem does this PR solve? 

Implement vectorized evaluation for builtinTidbParseTsoSig.

### What is changed and how it works?

```
BenchmarkVectorizedBuiltinTimeFunc/builtinTidbParseTsoSig-VecBuiltinFunc-4                  6111            186963 ns/op               0 B/op          0 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinTidbParseTsoSig-NonVecBuiltinFunc-4               5262            216277 ns/op               0 B/op          0 allocs/op 
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_time_vec.go [hunk 1]
```diff
@@ -816,11 +817,40 @@ func (b *builtinSubDateIntStringSig) vecEvalTime(input *chunk.Chunk, result *chu
...
 func (b *builtinTidbParseTsoSig) vecEvalTime(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETInt, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf); err != nil {
+		return handleInvalidTimeError(b.ctx, err)
```
Comment (SunRunAway) [inline expression/builtin_time_vec.go]: It's the `args[0]`'s responsibility to handle invalid time error, so we do not need handle it here.

File: expression/builtin_time_vec.go [hunk 2]
```diff
@@ -816,11 +817,40 @@ func (b *builtinSubDateIntStringSig) vecEvalTime(input *chunk.Chunk, result *chu
...
+			continue
+		}
+		t := oracle.GetTimeFromTS(uint64(args[i]))
+		r := types.Time{
+			Time: types.FromGoTime(t),
+			Type: mysql.TypeDatetime,
+			Fsp:  types.MaxFsp,
+		}
+		if err := r.ConvertTimeZone(time.Local, b.ctx.GetSessionVars().Location()); err != nil {
+			result.SetNull(i, true)
```
Comment (SunRunAway) [inline expression/builtin_time_vec.go]: This line is not needed since we return an error here.

File: expression/builtin_time_vec.go [hunk 3]
```diff
@@ -816,11 +817,39 @@ func (b *builtinSubDateIntStringSig) vecEvalTime(input *chunk.Chunk, result *chu
...
+	}
+	args := buf.Int64s()
+	result.ResizeTime(n, false)
+	result.MergeNulls(buf)
+	times := result.Times()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		t := oracle.GetTimeFromTS(uint64(args[i]))
```
Comment (Reminiscent) [inline expression/builtin_time_vec.go]: We need to check ```if args[i] < 0 ``` first.

File: expression/builtin_time_vec.go [hunk 4]
```diff
@@ -816,11 +817,39 @@ func (b *builtinSubDateIntStringSig) vecEvalTime(input *chunk.Chunk, result *chu
...
+		return err
+	}
+	args := buf.Int64s()
+	result.ResizeTime(n, false)
+	result.MergeNulls(buf)
+	times := result.Times()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
```
Comment (Reminiscent) [inline expression/builtin_time_vec.go]: We need to check ```if args[i] < 0 ``` first.

Comment (AerysNan) [reply expression/builtin_time_vec.go]: Updated, just set it to null if `args[i]` is less than or equal to zero.

**Comment by qw4990** [comment]
@AerysNan Please fix the CI problem:
```
[2019-11-20T13:19:21.169Z] gofmt (simplify)
[2019-11-20T13:19:27.787Z] expression/builtin_time_vec_test.go
```

It seems you should `gofmt -w` this file.

**Comment by AerysNan** [comment]
@qw4990 Thanks for your reminder, formatted.

**Pull request merged**